Overwatch TGUI

harryob opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

converts OW to using tgui from HTML

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding, and may discourage maintainers from reviewing or merging your PR. This section is not strictly required for (non-controversial) fix PRs or backend PRs. -->

soon the scourge that is HTML interfaces will be purged
<img width="906" alt="CleanShot 2022-10-12 at 08 34 10@2x" src="https://user-images.githubusercontent.com/55142896/195279256-0df5e073-af7a-43b5-b1a7-0569ef39610c.png">


## Changelog

<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: HarryOB
refactor: overwatch now uses TGUI
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +489,384 @@
+	"alive" = alive_squad_members,
+	)
+
+	return(squad_data)
```
```suggestion
	return squad_data
```
more readable

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +489,384 @@
+			"SSD" = !current_squaddie.key || !current_squaddie.client && current_squaddie.stat != DEAD
+			)
+
+	return(marines_list)
```
```suggestion
	return marines_list
```
same

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +489,384 @@
+	data["marine_filter"] = marine_filter
+	data["marine_filter_enabled"] = marine_filter_enabled
+	data["operator"] = operator?.name
+	data["bombardment_cooldown"] = COOLDOWN_TIMELEFT(almayer_orbital_cannon, ob_firing_cooldown)
```
feed world.time and ob_firing_cooldown into the UI then do the checks in the UI itself rather than in the DMcode

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +489,384 @@
+	data["almayer_cannon_disabled"] = almayer_orbital_cannon.is_disabled ? TRUE : FALSE
+
+	if(current_squad)
+		data["supply_cooldown"] = COOLDOWN_TIMELEFT(current_squad, next_supplydrop)
```
same

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +489,384 @@
+			return TRUE
+
+		if("pick_squad")
+			if(operator != usr)
+				return FALSE
```
do these checks in the UI, if they're already done in the UI you do not need them here

mullenpaul commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +489,384 @@
+			change_lead(params["marine_picked"])
+			return TRUE
+
+		if("insubordination")
```
Is there a way to remove insubordination?

harryob replied:
as it currently exists, it marks them for arrest on security terminals, which'd require using a security terminal to remove

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+      height={windowHeight}
+      scrollable
+      theme="weyland">
+      <Window.Content>{body}</Window.Content>
```
I would recommend using conditional rendering rather than using a variable.

`{operator && <OverwatchMain />}`
`{!operator && <OverwatchEmpty />}`

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+                  </Stack.Item>
+                  <Stack.Item grow>
+                    <Stack>
+                      {currentControlCategory === 1 ? (
```
Again, I would just fall back on conditional rendering - it makes this a bit easier to read.

`{(currentControlCategory  === 1) && ...}`
`{(currentControlCategory != 1) && ...}`

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+  return (
+    <>
+      <OverwatchId />
+      {squad_data ? (
```
Conditional rendering again. You have a fair bit of component code which is in this conditional segment. By the time you get to the false condition statement you need to scroll up to find out what the condition is.

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+        <Button
+          fluid
+          selected={currentControlCategory === 1}
+          textColor={currentControlCategory === 1 ? '#ffffff' : '#161613'}
```
I would recommend using the style sheets to style your components. It gives you a nice separation between component definition and styling.

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+          fluid
+          selected={currentControlCategory === 2}
+          icon="burst"
+          textColor={currentControlCategory === 2 ? '#ffffff' : '#161613'}
```
Ditto above regarding styling

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+            icon={'user'}
+            fluid
+            content={
+              squad_data ? 'Squad: ' + squad_data.squad_name : 'Squad: None'
```
You could use null coalescence here for something like

`'Squad: ' + squad_data?.squad_name ?? 'None'`

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+              <Table.Cell>Alive</Table.Cell>
+              <Table.Cell>Deployed</Table.Cell>
+            </Table.Row>
+            <Table.Row>
```
Why not have a single cell, which spans all rows and contains a divider? Alternatively, style the above row to have a border on the bottom

mullenpaul commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,860 @@
+        minValue={0}
+        maxValue={500}
+        value={cooldown}>
+        {cooldown + ' seconds'}
```
you don't need to cat the strings together, seconds can be outside of the braces

stanalbatross commented:
squad monitor section does not expand when you pull down the UI 😢

stanalbatross commented:
I would also condense the select squad and squad buttons into the same button

stanalbatross commented:
you should disable cam buttons of marines unable to be watched

stanalbatross commented:
sending squad and SL messages broken

stanalbatross commented:
needs to use the STRICT incapacitated and adjacent checker and also to reset view on close

stanalbatross commented:
![image](https://user-images.githubusercontent.com/66756236/194933679-dee4ea2d-0ee7-4b0d-b3be-6a0a385ce470.png)
 not descriptive enough, make it say "ready to fire" when there's 0 cooldown and make it say cooldown or w/e

stanalbatross commented:
feature idea : filter by job

stanalbatross commented:
![image](https://user-images.githubusercontent.com/66756236/194934616-84646033-db41-482d-8fbc-dc8a8c40bcaf.png)
column width should be a bit wider

stanalbatross commented:
merge squad select and operator control into one bar at the top to condense the UI

Segrain commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +489,388 @@
+	data["marine_filter"] = marine_filter
+	data["marine_filter_enabled"] = marine_filter_enabled
+	data["operator"] = operator?.name
+	data["bombardment_cooldown"] = COOLDOWN_TIMELEFT(almayer_orbital_cannon, ob_firing_cooldown)
```
![image](https://user-images.githubusercontent.com/4447185/194940231-b55bdf20-7fb1-4928-bfd9-7cfa1d079bce.png)
Console breaks on WO.

harryob replied:
of course it does

harryob replied:
fixed, sorry WO round

stanalbatross reviewed (changes requested):
runtimes : 
`[21:47:26]RUNTIME: Cannot read null.z - overwatch.dm@551<br>`
`[21:36:23]RUNTIME: Cannot execute null.release overwatch(). - overwatch.dm@666<br>`

harryob commented:
it's better now:
<img width="906" alt="CleanShot 2022-10-12 at 08 34 10@2x" src="https://user-images.githubusercontent.com/55142896/195279144-6138e8b9-22c9-4581-b5a6-5e2ee7f38e7b.png">


> squad monitor section does not expand when you pull down the UI 😢

fixed

> I would also condense the select squad and squad buttons into the same button

fixed

> you should disable cam buttons of marines unable to be watched

fixed

> sending squad and SL messages broken

not anymore

> needs to use the STRICT incapacitated and adjacent checker and also to reset view on close

thank god you added that

> squad transfers also broken

fixed

> ![image](https://user-images.githubusercontent.com/66756236/194933679-dee4ea2d-0ee7-4b0d-b3be-6a0a385ce470.png)
> not descriptive enough, make it say "ready to fire" when there's 0 cooldown and make it say cooldown or w/e

it now does

> merge squad select and operator control into one bar at the top to condense the UI

fixed

morrowwolf commented:
Would be nice if the screen saved how you resized it.

morrowwolf commented:
Tacmap popping up at random after checking it

morrowwolf commented:
Using the button in squad monitor cleanly switches screens but the squad leader button does not when switching squad leader to squad leader cams.

Please change the dialogue of "Latitude is now -397, longitude is now 199." to long and lat as is standard. Gave me a fucking hard attack.

morrowwolf commented:
Sort squad monitor by rank, more senior should be top. Or make option to sort per category.

fira reviewed (approved):
Would have been a good time to unscrew the computer having every piece of logic, but oh well

DM code LGTM

harryob commented:
> Would have been a good time to unscrew the computer having every piece of logic, but oh well
> 
> DM code LGTM

it's a worthy refactor for later, i'll look at doing it after this PR is done

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +473,343 @@
+	data["user"] = user
+
+	var/list/squad_list = list()
+	for(var/datum/squad/S in RoleAuthority.squads)
```
/S is le bad var name

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +473,343 @@
+		return
+
+	var/mob/user = usr
+	if(get_dist(user, src) > 1)
+		return FALSE
```
useless as ui_state etc already handles this

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +473,343 @@
+					current_squad.send_squad_message("Attention. [operator.name] has released overwatch system control. Overwatch functions deactivated.", displayed_icon = src)
+					to_chat(usr, "[icon2html(src, usr)] [SPAN_BOLDNOTICE("Overwatch system control override disengaged.")]")
+				else
+					var/mob/living/carbon/human/H = operator
```
var name

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +473,343 @@
+
+		if("pick_squad")
+			if(current_squad?.release_overwatch())
+				var/mob/living/carbon/human/H = operator
```
var name

stanalbatross commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -1065,6 +473,343 @@
+				var/squaddie = params["squaddie"]
+				if(!(squaddie in marine_filter))
+					marine_filter += squaddie
+					to_chat(usr, "[icon2html(src, usr)] [SPAN_NOTICE("Marine now hidden.")]")
+				else
+					marine_filter -= squaddie
+					to_chat(usr, "[icon2html(src, usr)] [SPAN_NOTICE("Marine will now be shown.")]")
```
to_chat should include the name of the marine hidden/shown I think

stanalbatross commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,809 @@
+          bad: [Infinity, 400],
+        }}
+        minValue={0}
+        maxValue={500}
```
this should not be hard coded as a number in the ui and should be got via static data instead because it is liable to change in the codebase

stanalbatross commented on tgui/packages/tgui/interfaces/OverwatchConsole.js:
```diff
@@ -0,0 +1,809 @@
+          bad: [Infinity, 400],
+        }}
+        minValue="0"
+        maxValue="500"
```
same here

sg2002 commented:
Two problems I've noticed in a test round:
1. When the DS gets locked, the OB says that it's not chambered. Probably need a separate message about the turret being locket.
2. After the hijack if you try to look at squad monitor it crashes for some of the squads. Couldn't narrow it further yet.
